instrumentation/elasticsearch/opentelemetry-instrumentation-elasticsearch.gemspec

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+lib = File.expand_path('lib', __dir__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require 'opentelemetry/instrumentation/elasticsearch/version'
+
+Gem::Specification.new do |spec|
+  spec.name        = 'opentelemetry-instrumentation-elasticsearch'
```

Comment:
This instrumentation library in its current state is only instrumenting [elasticsearch-transport](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport), which is a dependency of [elasticsearch](https://github.com/elastic/elasticsearch-ruby/blob/master/elasticsearch/elasticsearch.gemspec#L48-L49).  

I'm a little conflicted on whether or not we should be more explicit in that fact, or because the elasticsearch gem seems to just be a [bundler of the various elasticsearch "sub" gems](https://github.com/elastic/elasticsearch-ruby/blob/master/elasticsearch/lib/elasticsearch.rb) we should just leave this as is.  I'm leaning a bit towards the latter at the moment. 

@open-telemetry/ruby-approvers any stronger opinions here?

Reply:
Well ... at first I agreed, but now I'm not so sure. After reading through both `elasticsearch-transport` and `elasticsearch-api` ... I'm wondering if we should be instrumenting `elasticsearch-api` instead?

The default HTTP implementations supported by `elasticsearch-transport` are actually [provided by Faraday](https://github.com/elastic/elasticsearch-ruby/blob/master/elasticsearch-transport/lib/elasticsearch/transport/client.rb#L165-L166), and we already have `opentelemetry-instrumentation-faraday`. I think this PR would actually duplicate many of those spans.

Reply:
>Well ... at first I agreed, but now I'm not so sure. After reading through both elasticsearch-transport and elasticsearch-api ... I'm wondering if we should be instrumenting elasticsearch-api instead?

I'm trying to figure out if this is a point in favour of just simply keeping this as a more generic "elasticsearch" instrumentation library to go the more focused route of `opentelemetry-instrumentation-elasticsearch_api` or something to that effect.

>The default HTTP implementations supported by elasticsearch-transport are actually provided by Faraday, and we already have opentelemetry-instrumentation-faraday. I think this PR would actually duplicate many of those spans.

It looks like it will select between these lower level http libs, which we also have instrumentation for (with the exception of Patron).

>For optimal performance, use a HTTP library which supports persistent ("keep-alive") connections, such as patron or Typhoeus. Just require the library (require 'patron') in your code, and it will be automatically used.

>Currently these libraries will be automatically detected and used:

* Patron
* Typhoeus
* HTTPClient
* Net::HTTP::Persistent

- [from the readme](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport#elasticsearchtransport)

Reply:
> I agree, I think it makes sense.
> I will update the PR to instrument `elasticsearch-api` instead 👍
> 
> Thanks!

@mmontalvo mind leaving it for the moment?  Don't want you to change this before we settle on the approach.  Renaming can be a bit tedious.

Reply:
Hey @mmontalvo sorry about the delayed response. 

Regarding the name of this instrumentation gem, let's keep it as it is.

Regarding what portion of `elasticsearch` should be instrumented.  I went through the `elasticsearch-api` and the `elasticsearch-transport` gem.  

* All of the elasticsearch-api operations are just thin wrappers around the `perform_request` method that this PR is currently instrumenting.  Here's an [example](https://github.com/elastic/elasticsearch-ruby/blob/master/elasticsearch-api/lib/elasticsearch/api/actions/shutdown/get_node.rb#L18-L54).  Ideally it would be nice if we could add instrumentation points here so we could generate spans with the name of the initiating operation (`elasticsearch.get_node` or something to that effect).  I don't think there's a reasonable way to do this, and is probably not worth pursuing if we have to patch the many many methods like this that exist.  

* The elasticsearch-transport gem on the other hand has a single convenient instrumentation point (as you have done in this PR), the issue from my perspective is that it is not providing a lot of additional information if you take into consideration that we have faraday instrumentation.

If we take a look at the information faraday captures [here](https://github.com/open-telemetry/opentelemetry-ruby/blob/2e25dce78876f45d6774f6dafd0d6e317318ae6b/instrumentation/faraday/test/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware_test.rb#L50-L52), we would likely end up with two very similar spans with similar timing.  

I think the approach we want to take here is similar to that of `koala` which is an api client gem for facebook.  https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/koala/lib/opentelemetry/instrumentation/koala/patches/instrumentation.rb

The remaining question is whether or not this gem should depend on *and* add faraday instrumentation as part of its install process.

Reply:
Hey @robertlaurin no worries!
All the questions around the PR required some investigation and agreement into which way to go 😁 

I will take a look at the `koala` instrumentation and plan to instrument `elasticsearch-api` with the points you mentioned.

Thanks!